admin/UpdateDatabasePrivileges.pl

```diff
@@ -0,0 +1,147 @@
+
+If `--nogrant` is specified, then it also revokes read privileges
+(USAGE/SELECT).
+
+Options are:
+        --database          Database to connect to (default: MAINTENANCE)
+        --primary-ro-role   Name of the primary READONLY role
+                            (default: musicbrainz_ro)
+        --other-ro-role     Name of another role allowed RO access to the
+                            database (may be specified multiple times)
+                            (default: caa_redirect, sir)
+        --[no]grant         Whether to GRANT or REVOKE USAGE/SELECT privileges
+                            (default: GRANT)
+    -h, --help              Show this help
+
+EOF
+
+GetOptions(
+    'database=s'        => \$database,
+    'primary-ro-role=s' => \$primary_ro_role,
+    'other-ro-role=s'   => \@other_ro_roles,
+    'grant!'            => \$grant_privileges,
+    'help|h'            => sub { print $help; exit },
+) or exit 2;
+print($help), exit 2 if @ARGV;
+
+my $c = MusicBrainz::Server::Context->create_script_context(
+    database => 'SYSTEM_' . $database,
+);
+my $sql = $c->sql;
+
+$sql->auto_commit;
+my $existing_role_names = $sql->select_single_column_array(
+    <<~'SQL',
+    SELECT rolname
+      FROM pg_roles
+     WHERE rolname = any(?)
+    SQL
+    [$primary_ro_role, @other_ro_roles],
+);
+my %existing_role_names = map { $_ => 1 } @$existing_role_names;
+
+if ($existing_role_names{$primary_ro_role}) {
+    my $quoted_primary_role = $sql->dbh->quote_identifier($primary_ro_role);
+    $sql->begin;
+    for my $schema (@FULL_SCHEMA_LIST) {
+        my $quoted_schema = $sql->dbh->quote_identifier($schema);
+        my $revoked_privileges;
+        if ($grant_privileges) {
+            $sql->do(<<~"SQL");
+                GRANT USAGE
+                   ON SCHEMA $quoted_schema
+                   TO $quoted_primary_role;
+                SQL
+            $revoked_privileges =
+                'INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER';
+        } else {
+            $sql->do(<<~"SQL");
+                REVOKE USAGE
+                    ON SCHEMA $quoted_schema
+                  FROM $quoted_primary_role;
+                SQL
+            $revoked_privileges = 'ALL PRIVILEGES';
+        }
+        $sql->do(<<~"SQL");
+            REVOKE $revoked_privileges
+                ON ALL TABLES IN SCHEMA $quoted_schema
+              FROM $quoted_primary_role;
+            SQL
+        $sql->do(<<~"SQL") if $grant_privileges;
+            GRANT SELECT
+               ON ALL TABLES IN SCHEMA $quoted_schema
+               TO $quoted_primary_role;
+            SQL
+    }
+    for my $other_role (@other_ro_roles) {
+        unless ($existing_role_names{$other_role}) {
+            log_info {
+                qq(The role "$other_role" does not exist; skipping.)
+            };
+            next;
+        }
+
+        my $quoted_other_role = $sql->dbh->quote_identifier($other_role);
```

Comment:
It would be helpful to log roles whose database privileges are being updated, and in which database.